Update script scaffold templates for config_flow_oauth2

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update the generated code produced by script.scaffold for config_flow_oauth2 integrations to:
1. Pass with fewer modifications
2. Match the Python API Library documentation

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This was tested by creating a new skeleton integration with ```python3 -m script.scaffold integration```, making a small modification to api.py to remove a placeholder dependency, and running the newly created tests.

- This PR fixes or closes issue: fixes # 41152
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Update scaffolding config_flow_oauth2 to pass

Update the code generated by script.scaffolding to add missing dependencies
on http, and initialization of current_request so that the tests pass with
fewer modifications.  The modification is based on other components that
use LocalOAuth2Implementation.

sanyatuning commented on script/scaffold/templates/config_flow_oauth2/integration/api.py:
```diff
@@ -35,7 +35,7 @@ def refresh_tokens(self) -> dict:
         ).result()
 
-        return self.session.token
+        return self.session.token["access_token"]
```
This is not a dictionary I think

allenporter replied:
My reading of OAuth2Session (which matches the behavior seen when debugging a new integration with this scaffold) is that it is a dict with fields {"expires_at", "access_token", "refresh_token", "expires_in"}.  Basically returning the entire json response all the way up from  LocalOAuth2Implementation._token_request

https://github.com/home-assistant/core/blob/0cce35b23edffd7beffe6b817fa7a357265bf0ca/homeassistant/helpers/config_entry_oauth2_flow.py#L430

I could see an argument that maybe LocalOAuth2Implementation should change its behavior to actually return a "token" here.  Is there another similar oauth session implementation somewhere with a different behavior?

balloob replied:
What @sanyatuning probably means is that the `access_token` value inside `session.token` is not a dictionary.

balloob replied:
The idea here was to just return the last token request from the server, so you can use both the token type and the access token for your request.

allenporter committed:
Update OAuth scaffold API to match tutorial

Pull out the "access_token" from the token dictionary to correctly implement
the AbstractAuth expectations.  This makes the default scaffolding match
the expectations of the examples in the Python API Library documentation.

balloob commented on script/scaffold/templates/config_flow_oauth2/integration/api.py:
```diff
@@ -55,4 +55,4 @@ async def async_get_access_token(self):
             await self._oauth_session.async_ensure_token_valid()
 
-        return self._oauth_session.token
+        return self._oauth_session.token["access_token"]
```
Now the async and sync implementations are no longer in sync. 

Can you explain why you want to make this change?

balloob replied:
Ah I see now. Yeah the documentation probably also should have an update. 

I think in this case let's indeed update both sync + async to return the actual access token (like you did here) 👍

balloob committed:
Fix sync sample.
